spec/rubocop/cop/lint/no_return_in_begin_end_blocks_spec.rb

```diff
@@ -0,0 +1,136 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::Lint::NoReturnInBeginEndBlocks do
+  subject(:cop) { described_class.new(config) }
+
+  let(:config) { RuboCop::Config.new }
+
+  def return_inside_or_assignment_block
+    <<-RUBY
+      @bad_method ||= begin
+        return "odd number" if rand(1..2).odd?
+
+        "even number"
+      end
+    RUBY
+  end
+
+  def no_return_inside_or_assignment_block
+    <<-RUBY
+      @good_method ||= begin
+        if rand(1..2).odd?
+          "odd number"
+        else
+          "even number"
+        end
+      end
+    RUBY
+  end
+
+  def return_inside_simple_assignment_block
+    <<-RUBY
+      bad_method = begin
+        return "odd number" if rand(1..2).odd?
+
+        "even number"
+      end
+    RUBY
+  end
+
+  def return_inside_add_and_assignment_block
+    <<-RUBY
+      some_value = 10
+
+      some_value += begin
+        return 1 if rand(1..2).odd?
+
+        2
+      end
+    RUBY
+  end
+
+  def return_inside_subtract_and_assignment_block
+    <<-RUBY
+      some_value = 10
+
+      some_value -= begin
+        return 1 if rand(1..2).odd?
+
+        2
+      end
+    RUBY
+  end
+
+  def return_inside_multiply_and_assignment_block
+    <<-RUBY
+      some_value = 10
+
+      some_value *= begin
+        return 1 if rand(1..2).odd?
+
+        2
+      end
+    RUBY
+  end
+
+  def return_inside_divide_and_assignment_block
+    <<-RUBY
+      some_value = 10
+
+      some_value /= begin
+        return 1 if rand(1..2).odd?
+
+        2
+      end
+    RUBY
+  end
+
+  def return_inside_exponent_and_assignment_block
+    <<-RUBY
+      some_value = 10
+
+      some_value **= begin
+        return 1 if rand(1..2).odd?
+
+        2
+      end
+    RUBY
+  end
+
+  shared_examples 'accepts return inside a block' do |name, code|
+    it "accepts a return statement inside a block when using #{name}" do
+      expect_no_offenses(send(code))
```

Comment:
👍  Moreover, these particular `shared_examples` are not needed at all and should be avoided altogether. I find it much harder to know what test is failing when shared examples are used.

Personally, I'm more lazy than @jonas054 and a single example of `*=` satisfies me, I wouldn't list all the `+=` and variants.

Reply:
Thank you both for the feedback. Yes, my initial implementation was not the best one. I really liked the way @jonas054 approached the implementation of the shared examples. As this issue will affect those operator assignments, I think it's valid to have this test for them. As they're not so many, in case of failures I believe it would be easy to spot what's going on. On a more complex scenario, perhaps I'd also avoid abusing the use of shared examples. But in this case I think it's fine :) 

I'm struggling at the moment to actually add an offense for the `=` (ivasgn) test case. Not sure how to detect the offense in this case as the only content for the node in this case only has `s(:ivasgn, :@good_method)` and nothing more.

Here's how I'm trying to detect the offense at the moment:

```ruby
def on_ivasgn(node)
  node.each_node(:kwbegin) do |kwbegin_node|
    kwbegin_node.each_node(:return) do |return_node|
      add_offense(return_node)
    end
  end
end
```

Based on this output from the AST parsed code:

```ruby
(def :bad_method_with_operator
  (args
    (arg :memoized)
    (arg :operator))
  (ivasgn :@bad_method
    (kwbegin
      (if
        (send
          (lvar :memoized) :!)
        (return
          (int 1)) nil)
      (int 2))))
```

Any advice you can share with me on how to solve this one?

Cheers.

Reply:
Alright. I think I managed to solve the issue. The issue was that the cop's code was not considering multiple nodes (as in the example above) and was also returning when the node was not of a `begin_type`. 

I've updated the cop's code to consider these cases now, and also created an alias for the other operations as the check will be the same. 

With these changes, all tests are passing now for all the operators.